### DISCUSSION TITLE
Genomebin matrix

### DISCUSSION THREAD

**Discussion by AnjaliC4** [stuart-lab/signac#639] [Q&A] [answered]
Hi, I am trying to create genomic bin matrix in Signac and adding it as another assay in signac object"
atac.bins <- GenomeBinMatrix(
  fragments = frags,
  genome = genome,
  binsize = 1000,
  process_n = 2000
)
However, even with 100GB memory and 32 CPUs per task, it throws me an out-of-memory error. Although, it has only 5000 cells.

I would like to add a genome bin assay for clustering of signac objects. Please let me know how can I run this faster/using less memory?

**Comment by timoast** [comment, marked as answer]
Some suggestions:

* set the `cells` parameter in `GenomeBinMatrix()`
* Set the parallelization parameter (`workers`) lower, if enabling paralellization
* If enabling parallelization, raise the `future.globals.maxSize` option (see [vignette](https://satijalab.org/signac/articles/future.html))
* choose a larger bin size (eg, 5000)

**Reply by AnjaliC4** [reply]
Hi, thanks for the suggestions. Regarding your 1st suggestion:
I've made a fragment object with the chosen (5000) cells for analysis. So, when I try the GenomeBinMatrix function with a vector of the same cell names which were used to create fragment object, it throws me this error:
(I guess it's redundant to specify cells if I've already created a fragment object with those cells?)

```r
atac.bins <- GenomeBinMatrix(
fragments = frags,
genome = genome,
cells = rownames(cells),
binsize = 1000,
process_n = 2000
)

Extracting reads overlapping genomic regions
Error in (function (cl, name, valueClass)  :
  assignment of an object of class "numeric" is not valid for @'Dim' in an object of class "dgTMatrix"; 
 is(value, "integer") is not TRUE
Calls: GenomeBinMatrix ... resolve.list -> signalConditionsASAP -> signalConditions
Execution halted
```

**Reply by timoast** [reply]
What is `cells` here?

**Reply by AnjaliC4** [reply]
head(cells)
                                                                     bc         total_frags   frac_mito  frac_peak
ACCTCCAACAGCCTAT ACCTCCAACAGCCTAT       27710 0.001948755 0.1411765
CAGAAACCTAAATCGA CAGAAACCTAAATCGA        5994 0.007841175 0.2590924
TCCGAGTCTGGATGGT TCCGAGTCTGGATGGT       14497 0.001379596 0.2177002
TGTACGATGAGTCGGA TGTACGATGAGTCGGA       83437 0.006855472 0.4860074
AAGAATGTGATCCTCC AAGAATGTGATCCTCC       15162 0.003627490 0.2512202
TCGCGTACTATACGCT TCGCGTACTATACGCT       16695 0.001437556 0.2911650

 I get this with rownames(cells) (which I pass to fragment object and GenomeBin matrix)
([1] "ACCTCCAACAGCCTAT" "CAGAAACCTAAATCGA" "TCCGAGTCTGGATGGT" "TGTACGATGAGTCGGA"
[5] "AAGAATGTGATCCTCC" "TCGCGTACTATACGCT"

Btw, the fragment object was created by passing this as cells.  I also tried without using the 'cells' option, interestingly, even by passing a fragment obj of subsetted cells in the genome matrix, the matrix was created with all the cells. I have pasted the code and errors here. (It seems like passing cells is important- but then it gives me the error pasted above)

frags <- CreateFragmentObject(
  path = "/atac/summary/atac.fragments.sorted.bed.gz",
  cells = rownames(cells)
)
Computing hash
Checking for 6309 cell barcodes
> frags
A Fragment object for 6309 cells

plan("multiprocess", workers = 5)
options(future.globals.maxSize = 60 * 1024 ^ 3)
plan()

genome <- seqlengths(BSgenome.Hsapiens.UCSC.hg38)
atac.bins <- GenomeBinMatrix(
  fragments = frags,
  genome = genome,
  binsize = 1000,
  process_n = 5000
)

dim(atac.bins)
[1] 3099814 6105920

atac_obj
An object of class Seurat
248823 features across 6309 samples within 1 assay
Active assay: peaks (248823 features, 0 variable features)
> head(colnames(atac_obj))
[1] "ACCTCCAACAGCCTAT" "CAGAAACCTAAATCGA" "TCCGAGTCTGGATGGT" "TGTACGATGAGTCGGA"
[5] "AAGAATGTGATCCTCC" "TCGCGTACTATACGCT"
> head(rownames(atac_obj))
[1] "chr1-9947-10471"    "chr1-180588-181088" "chr1-181220-181720"
[4] "chr1-191050-191928" "chr1-605219-605719" "chr1-629468-630366"

atac_obj[['bins']] <- CreateAssayObject(counts = atac.bins)

Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
Error: Cannot add a different number of cells than already present
Execution halted

**Reply by timoast** [reply]
>  interestingly, even by passing a fragment obj of subsetted cells in the genome matrix, the matrix was created with all the cells

This is expected, if the `cells` parameter is not set in `GenomeBinMatrix()` then the resulting matrix will contain every barcode found in the fragment file (_not_ the `Fragment` object). See documentation for [`GenomeBinMatrix`](https://satijalab.org/signac/reference/genomebinmatrix)

The second error you have is because you're trying to add an assay containing 6105920 cells to a Seurat object containing 6309 cells. If you first subset the genome bin matrix to only contain cells in the object, you should not see this error.

For this other error:

```r
atac.bins <- GenomeBinMatrix(
fragments = frags,
genome = genome,
cells = rownames(cells),
binsize = 1000,
process_n = 2000
)

Extracting reads overlapping genomic regions
Error in (function (cl, name, valueClass)  :
  assignment of an object of class "numeric" is not valid for @'Dim' in an object of class "dgTMatrix"; 
 is(value, "integer") is not TRUE
Calls: GenomeBinMatrix ... resolve.list -> signalConditionsASAP -> signalConditions
Execution halted
```

Can you open a new issue containing the full code and output of `sessionInfo()`, and if possible a reproducible example using a public dataset?

**Reply by AnjaliC4** [reply]
Yes, I will do that. Thanks!
